regex/src/main/scala/regex/Regex.scala

```diff
@@ -15,14 +29,53 @@ package regex
   */
 
 /** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+def simplify(lang: RegularLanguage): RegularLanguage = 
+  lang match {
+    case Concat(Epsilon, input) => simplify(input)
+    case Concat(input, Epsilon) => simplify(input)
+    case Concat(Empty, input) => Empty
+    case Concat(input, Empty) => Empty
+    case Concat(input1, input2) => Concat(simplify(input1), simplify(input2))
+    case Union(Empty, input) => simplify(input)
+    case Union(input, Empty) => simplify(input)
+    case Union(input1, input2) => Union(simplify(input1), simplify(input2))
+    case Star(Epsilon) => Epsilon
+    case Star(Empty) => Empty
+    case Star(input) => Star(simplify(input))
+    case _ => lang
+  } 
 
 /** A language is nullable if it contains ε */
-def nullable(lang: RegularLanguage): Boolean = ???
+def nullable(lang: RegularLanguage): Boolean = 
+  val simplified = simplify(lang)
+  simplified match{
+  // lang match{
+    case Empty => false
+    case Epsilon => true
+    case Character(someChar) => false
+    case Concat(input1, input2) => nullable(input1) && nullable(input2)
+    case Union(input1, input2) => nullable(input1) || nullable(input2)
+    case Star(input) => true
+  }
 
 /** Computes the derivative of a language, with respect to a character */
-def derivative(l: RegularLanguage)(c: Char): RegularLanguage = ???
-
+def derivative(l: RegularLanguage)(c: Char): RegularLanguage = 
```

Comment:
I like how you indented the lines for clarity and to follow the 80-character-per-line convention (I forgot to do that for my solution…). Another way that both of us could take to make the code more readable is by defining more temporary variables, like in the sample solution. 

While doing this part, I found “pattern guards” that allow writing the if statement with the case. Here is a link to the documentation if you are interested: https://docs.scala-lang.org/tour/pattern-matching.html